lib/kube/proxy/forwarder.go

```diff
@@ -453,7 +453,14 @@ func (f *Forwarder) withAuth(handler handlerWithAuthFunc) httprouter.Handle {
 		if err := f.authorize(req.Context(), authContext); err != nil {
 			return nil, trace.Wrap(err)
 		}
-		if err := f.acquireConnectionLock(req.Context(), authContext); err != nil {
+
+		user := authContext.Identity.GetIdentity().Username
+		roles, err := getRolesByName(f, authContext.Identity.GetIdentity().Groups)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if err := f.AcquireConnectionLock(req.Context(), user, roles); err != nil {
```

Comment:
Where is this semaphore released btw? I don't see if in the code unless I'm missing something.

Reply:
Ah, they're all cancelled when the `req.Context()` cancels. This happens whenever the request gets completed/the socket closes.

Reply:
```
// AcquireSemaphoreLock attempts to acquire and hold a semaphore lease.  If successfully acquired,
// background keepalive processes are started and an associated lock handle is returned.  Cancelling
// the supplied context releases the semaphore.```
as per the semaphore godoc

Reply:
Got it!